Remove MPI from serial ADIOS interface

anokfireball opened the pull request:
To be able to compile both serial and parallel ADIOS1 backends, remove
all traces of MPI from the serial interface. MPI_COMM_NULL is used in
the implementation, but not exposed.

Suggested in https://github.com/openPMD/openPMD-api/issues/252#issuecomment-396907630, blocker for #254.

anokfireball committed:
Remove MPI from serial ADIOS interface

To be able to compile both serial and parallel ADIOS1 backends, remove
all traces of MPI from the serial interface. MPI_COMM_NULL is used in
the implementation, but not exposed.

anokfireball commented on test/SerialIOTest.cpp:
```diff
@@ -1121,7 +1121,7 @@ TEST_CASE( "no_serial_hdf5", "[serial][hdf5]" )
 }
 #endif
-#if openPMD_HAVE_ADIOS1
+#if openPMD_HAVE_ADIOS1 && !openPMD_HAVE_MPI
```
You can revert this change when #254 is ready.

ax3l replied:
uff, I rebased on your branch and applied the mentioned changes.
I won't get a running solution with our new approach unless I do explicit symbol hiding/exposing in the two transalation units for adios, since otherwise symbols like `adios_open` and `common_adios_open` will be mismatched at link and runtime.

Can do, but this will exclude pgic++ at least since not all compilers know symbol hiding. harsh world...

Or so some symbol renaming with `--prefix-symbols` via `objcpy`... gosh!

ax3l replied:
We could maybe go for `dlopen()` for one ADIOS part xD

ax3l commented on src/IO/ADIOS/ParallelADIOS1IOHandler.cpp:
```diff
@@ -36,74 +37,59 @@ namespace openPMD
 {
-    char const* env = std::getenv(key.c_str());
-    if( env != nullptr )
+    /* create all files where ADIOS file creation has been deferred, but this has never been triggered
```
the comment is not very clear: what is "this" what has never been triggered and what is "this" that happens in the line below?

anokfireball replied:
True, the sentence really is a bit vague.
First "this" means actual file creation. The operation (file creation) has been deferred, but because no IO op that triggers execution of the deferred operation (file creation) has been scheduled across the whole Series, we have to do it manually.
Seconds "this" stands for the situation described above (i.e. the scenario in which no appropiate IO op has been scheduled).

ax3l commented:
really nice, thank you!

ax3l commented on include/openPMD/IO/ADIOS/ADIOS1IOHandler.hpp:
```diff
@@ -25,7 +25,6 @@
 #if openPMD_HAVE_ADIOS1
 #   include "openPMD/IO/AbstractIOHandlerImpl.hpp"
 #   include <adios.h>
```
hm, `adios.h` also includes `adios_mpi.h`... argh.. but this calls the dummy... of have to check my setters more.

ax3l replied:
I think you need `adios.h` and `mpi.h` in `ParallelADIOS1...hpp` only.

In the serial header, use `adios_read.h` and in the serial impl. `mpidummy.h` as you do.

anokfireball committed:
Pull adios.h into implementation

The adios.h header is not required for the interfaces.
This might make separate compilation and multiple linking easier.
